# Puzzle 1

¡Estás ocupándote de tus propios asuntos en un barco en el mar cuando suena la alarma de mar! Te apresuras a ver si puedes ayudar. ¡Aparentemente, uno de los Elfos tropezó y accidentalmente envió las llaves del trineo volando al océano!

Antes de que te des cuenta, estás dentro de un submarino que los Elfos mantienen preparados para situaciones como esta. Está cubierto de luces navideñas (porque por supuesto que lo está), e incluso tiene una antena experimental que debería poder rastrear las teclas si puede aumentar la intensidad de la señal lo suficiente; hay un pequeño medidor que indica la intensidad de la señal de la antena mostrando 0-50 estrellas.

Tus instintos te dicen que para salvar la Navidad, necesitarás obtener las cincuenta estrellas antes del 25 de diciembre.

Recoge estrellas resolviendo acertijos. Dos rompecabezas estarán disponibles cada día en el calendario de Adviento; el segundo rompecabezas se desbloquea cuando completas el primero. Cada rompecabezas otorga una estrella. ¡Buena suerte!

A medida que el submarino desciende por debajo de la superficie del océano, automáticamente realiza un barrido de sonar del fondo marino cercano. En una pantalla pequeña, aparece el informe de barrido de la sonda (la entrada del rompecabezas): cada línea es una medida de la profundidad del fondo del mar a medida que el barrido se ve cada vez más lejos del submarino.

Por ejemplo, suponga que tiene el siguiente informe:

199
200
208
210
200
207
240
269
260
263
Este informe indica que, escaneando hacia afuera desde el submarino, el barrido del sonar encontró profundidades de 199, 200, 208, 210, etc.

La primera orden del día es averiguar qué tan rápido aumenta la profundidad, solo para que sepa con qué está lidiando: nunca se sabe si las llaves serán llevadas a aguas más profundas por una corriente oceánica o un pez o algo así.

Para hacer esto, cuente el número de veces que una medición de profundidad aumenta con respecto a la medición anterior. (No hay medición antes de la primera medición). En el ejemplo anterior, los cambios son los siguientes:

199 (N / A - sin medición previa)
200 (aumentado)
208 (aumentado)
210 (aumentado)
200 (disminuido)
207 (aumentado)
240 (aumentado)
269 ​​(aumentado)
260 (disminuido)
263 (aumentado)
En este ejemplo, hay 7 medidas que son más grandes que la medida anterior.

¿Cuántas medidas son más grandes que la medida anterior?

## Logica de Resolucion


Bueno Para comenzar lo que nos piden es basicamente encontrar los numeros que sonn mayores a los numeros que le preceden, para eso:

*  Hay que saber que al primer numero no lo precede ningun numero, es decir vamos a tomar ese como nuestro primer valor y lo dejaremos guardado en una variable llamada `valor`.

* Luego en una variable llamada `contador` almacenaremos la cantidad de veces que encontramos que un numero es mas grande que el numero que lo precede.

* Con un ciclo for vamos a recorrer todos los datos dentro del archivo, linea por linea, y guadaremos cada valor en una variable llamada `valor_actual`

* Luego simplemente preguntamos, si `valor_actual` es mayor que `valor` y si es asi incrementamos en uno el valor de la variable `contador`, ya que esto significa que encontramos uno de los daros que necesitabamos !

* Vamos a ir actualizando `valor` con lo que estaba almacenado en `variable_actual` para siempre compararlo con el anterior

* Mostramos el dato almacenado en la variable `contador` y eso nos dara nuestro Resultado :)

## Entrada de Datos

Si quieres ver la entrada de datos que me toco para practicar, se encuentra en el archivo `input_puzzle1.txt`

## Codigo

In [43]:
contenido = open('input_puzzle1.txt', 'r')
mensaje = contenido.readlines()
valor = int(mensaje[0])
contador = 0
for linea in mensaje:
    valor_nuevo = int(linea)
    if valor_nuevo > valor:
        contador += 1
    valor = valor_nuevo
print(contador)
    

1616


## Resultado Parte1: 1616

## Parte 2

--- La segunda parte ---
Tener en cuenta que cada medición no es tan útil como esperaba: hay demasiado ruido en los datos.

En su lugar, considere las sumas de una ventana deslizante de tres medidas. Nuevamente considerando el ejemplo anterior:

199 A
200 A B
208 A B C
210 B C D
200 E C D
207 E F D
240 E F G
269 ​​F G H
260 G H
263 H
Empiece por comparar la primera y la segunda ventana de tres mediciones. Las medidas de la primera ventana están marcadas con A (199, 200, 208); su suma es 199 + 200 + 208 = 607. La segunda ventana está marcada como B (200, 208, 210); su suma es 618. La suma de las medidas en la segunda ventana es mayor que la suma de la primera, por lo que esta primera comparación aumentó.

Su objetivo ahora es contar el número de veces que la suma de medidas en esta ventana deslizante aumenta con respecto a la suma anterior. Entonces, compare A con B, luego compare B con C, luego C con D, y así sucesivamente. Deténgase cuando no queden suficientes medidas para crear una nueva suma de tres medidas.

En el ejemplo anterior, la suma de cada ventana de tres medidas es la siguiente:

A: 607 (N / A - sin suma previa)
B: 618 (aumentado)
C: 618 (sin cambios)
D: 617 (disminuido)
E: 647 (aumentado)
F: 716 (aumentado)
G: 769 (aumentado)
H: 792 (aumentado)
En este ejemplo, hay 5 sumas que son mayores que la suma anterior.

Considere las sumas de una ventana deslizante de tres medidas. ¿Cuántas sumas son mayores que la suma anterior?

## Logica de Resolucion

Utilizando la misma entrada de datos, ahora lo que nos piden es separar eso datos en grupos de a 3, y ver si la suma de los grupos (en este ejercicio le llaman ventanas) es mayor que el grupo anterior, para eso:

* Uutilizamos una lista para guardar los valores y luego sumamos los valores dentro de la lista para luego compararala.

* en una variable llamada `limite` guardamos el maximo valor hasta el cual va a iterar el ciclo for, ya que si se pasa de esevalor no va a poder armar los grupos de 3.

* con cun ciclo recorremos los datos y si no estamos iterando por encima del `limite` entonces, vamos a capturar nuevos grupos de 3, sumar los valores dentro del grupo y compararlos con los del grupo anterior.

* en `contador` guardaremos el valor de la cantidad de grupos, que su suma de valores, es mayor que el grupo anterior, entonces, cada vez que encuntre algo asi se incrementara en 1.

* actualizamos el valor de `suma`con `suma_actual` para siempre ir comparando con el ultimo
* Al final mostraremos el valor almacenado en `contador` y eso nos dara el resultado



## Codigo

In [41]:
contenido = open('input_puzzle1.txt', 'r')
mensaje = contenido.readlines()
j = 0
lista_inicial = [int(str(mensaje[j]).strip()), 
                 int(str(mensaje[j + 1]).strip()),
                 int(str(mensaje[j + 2]).strip())]
suma = sum(lista_inicial)
limite = len(mensaje) - 2
contador = 0
for i in range(len(mensaje)):
    if (i < limite):
        lista = [int(str(mensaje[i]).strip()), 
                int(str(mensaje[i + 1]).strip()), 
                int(str(mensaje[i + 2]).strip())]
        suma_actual = sum(lista)
        if suma_actual > suma:
            contador += 1
        suma = suma_actual 
print(f' contador: {contador}')


 contador: 1645


## Respuesta Parte2: 1645